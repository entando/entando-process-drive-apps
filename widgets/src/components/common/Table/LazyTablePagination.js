import PropTypes from 'prop-types';
import React from 'react';
import PlayArrowIcon from '@material-ui/icons/PlayArrow';
import SkipPreviousIcon from '@material-ui/icons/SkipPrevious';
import withStyles from '@material-ui/core/styles/withStyles';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import TableCell from '@material-ui/core/TableCell';
import i18next from 'i18next';

const styles = {
  root: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  pageControl: {
    display: 'flex',
    alignItems: 'center',
  },
  previousIcon: {
    transform: 'rotate(180deg)',
  },
  select: {
    textAlign: 'right',
    textAlignLast: 'right',
    marginLeft: 8,
    fontSize: 'inherit',
    '&:before, &:hover:before': {
      content: 'none',
    },
  },
};

class LazyTablePagination extends React.Component {
  handleFirstPageButtonClick = () => {
    const { onChangePage } = this.props;
    onChangePage(0);
  };

  handlePrevButtonClick = () => {
    const { onChangePage, page } = this.props;
    onChangePage(page - 1);
  };

  handleNextButtonClick = () => {
    const { onChangePage, page } = this.props;
    onChangePage(page + 1);
  };

  handleRowsPerPage = e => {
    const { onChangeRowsPerPage } = this.props;
    onChangeRowsPerPage(e);
  };

  render() {
    const { page, rowsPerPage, rowsPerPageOptions, lastPage, classes } = this.props;

    return (
      <TableCell className={classes.root}>
        <div className={classes.pageControl}>
          <Typography variant="caption">{i18next.t('table.rowsPerPage')}</Typography>
          <Select value={rowsPerPage} className={classes.select} onChange={this.handleRowsPerPage}>
            {rowsPerPageOptions.map(item => (
              <MenuItem key={item} value={item}>
                {item}
              </MenuItem>
            ))}
          </Select>
        </div>
        <div className={classes.pageControl}>
          <IconButton size="small" onClick={this.handleFirstPageButtonClick} disabled={page === 0}>
            <SkipPreviousIcon />
          </IconButton>
          <IconButton size="small" onClick={this.handlePrevButtonClick} disabled={page === 0}>
            <PlayArrowIcon className={classes.previousIcon} />
          </IconButton>

          <Typography>{page + 1}</Typography>

          <IconButton size="small" onClick={this.handleNextButtonClick} disabled={lastPage}>
            <PlayArrowIcon />
          </IconButton>
        </div>
      </TableCell>
    );
  }
}

LazyTablePagination.propTypes = {
  classes: PropTypes.shape({
    root: PropTypes.string,
    previousIcon: PropTypes.string,
    pageControl: PropTypes.string,
    select: PropTypes.string,
  }),
  onChangePage: PropTypes.func.isRequired,
  page: PropTypes.number.isRequired,
  rowsPerPage: PropTypes.number.isRequired,
  rowsPerPageOptions: PropTypes.arrayOf(PropTypes.number).isRequired,
  lastPage: PropTypes.bool.isRequired,
  onChangeRowsPerPage: PropTypes.func.isRequired,
};

LazyTablePagination.defaultProps = {
  classes: {},
};

export default withStyles(styles)(LazyTablePagination);
