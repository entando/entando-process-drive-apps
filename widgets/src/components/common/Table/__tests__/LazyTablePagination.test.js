import React from 'react';
import { cleanup, render } from '@testing-library/react';

import LazyTablePagination from 'components/common/Table/LazyTablePagination';

const rowsPerPageOptions = [5, 10, 15];

// eslint-disable-next-line react/prop-types
const TableHolder = ({ children }) => (
  <table>
    <tbody>
      <tr>{children}</tr>
    </tbody>
  </table>
);

describe('<LazyTablePagination />', () => {
  afterEach(cleanup);

  it('renders snapshot correctly', () => {
    const { container } = render(
      <TableHolder>
        <LazyTablePagination
          page={0}
          rowsPerPage={5}
          onChangePage={jest.fn()}
          onChangeRowsPerPage={jest.fn()}
          rowsPerPageOptions={rowsPerPageOptions}
          lastPage={false}
        />
      </TableHolder>
    );

    expect(container).toMatchSnapshot();
  });

  it('check for pagination buttons when page === 0', () => {
    const { container } = render(
      <TableHolder>
        <LazyTablePagination
          page={0}
          rowsPerPage={5}
          onChangePage={jest.fn()}
          onChangeRowsPerPage={jest.fn()}
          rowsPerPageOptions={rowsPerPageOptions}
          lastPage={false}
        />
      </TableHolder>
    );

    const buttons = container.querySelectorAll('button');
    expect(buttons[0].hasAttribute('disabled')).toBe(true);
    expect(buttons[1].hasAttribute('disabled')).toBe(true);
    expect(buttons[2].hasAttribute('disabled')).toBe(false);
  });

  it('check for pagination buttons when page is last', () => {
    const { container } = render(
      <TableHolder>
        <LazyTablePagination
          page={3}
          rowsPerPage={5}
          onChangePage={jest.fn()}
          onChangeRowsPerPage={jest.fn()}
          rowsPerPageOptions={rowsPerPageOptions}
          lastPage
        />
      </TableHolder>
    );

    const buttons = container.querySelectorAll('button');
    expect(buttons[0].hasAttribute('disabled')).toBe(false);
    expect(buttons[1].hasAttribute('disabled')).toBe(false);
    expect(buttons[2].hasAttribute('disabled')).toBe(true);
  });
});
