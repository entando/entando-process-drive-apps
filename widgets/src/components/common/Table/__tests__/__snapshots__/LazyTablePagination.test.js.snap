// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LazyTablePagination /> renders snapshot correctly 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="MuiTableCell-root LazyTablePagination-root-1"
        >
          <div
            class="LazyTablePagination-pageControl-2"
          >
            <span
              class="MuiTypography-root MuiTypography-caption"
            />
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline LazyTablePagination-select-4"
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby=" "
                class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                role="button"
                tabindex="0"
              >
                5
              </div>
              <input
                type="hidden"
                value="5"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
          <div
            class="LazyTablePagination-pageControl-2"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
                  />
                </svg>
              </span>
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root LazyTablePagination-previousIcon-3"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8 5v14l11-7z"
                  />
                </svg>
              </span>
            </button>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              1
            </p>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8 5v14l11-7z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
